ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/LedgerConfigProvisioner.scala

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.apiserver.configuration
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import akka.stream.Materializer
+import com.daml.api.util.TimeProvider
+import com.daml.ledger.api.SubmissionIdGenerator
+import com.daml.ledger.participant.state.{v2 => state}
+import com.daml.ledger.resources.ResourceOwner
+import com.daml.lf.data.Time.Timestamp
+import com.daml.logging.LoggingContext.withEnrichedLoggingContext
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.configuration.LedgerConfiguration
+import com.daml.telemetry.{DefaultTelemetry, SpanKind, SpanName}
+
+import scala.compat.java8.FutureConverters
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationLong
+
+/** Writes a default ledger configuration to the ledger, after a configurable delay. The
+  * configuration is only written if the ledger does not already have a configuration.
+  *
+  * Used by the participant to initialize a new ledger.
+  */
+private[apiserver] final class LedgerConfigProvisioner private (
+    currentLedgerConfiguration: CurrentLedgerConfiguration,
+    writeService: state.WriteConfigService,
+    timeProvider: TimeProvider,
+    submissionIdGenerator: SubmissionIdGenerator,
+    config: LedgerConfiguration,
+    materializer: Materializer,
+)(implicit loggingContext: LoggingContext)
+    extends AutoCloseable {
+  private val logger = ContextualizedLogger.get(getClass)
+
+  private val closed: AtomicBoolean = new AtomicBoolean(false)
+
+  materializer.scheduleOnce(
+    config.initialConfigurationSubmitDelay.toNanos.nanos,
+    () => {
+      if (currentLedgerConfiguration.latestConfiguration.isEmpty && !closed.get)
+        submitInitialConfig(writeService)
+      ()
+    },
+  )
+
+  // There are several reasons why the change could be rejected:
+  // - The participant is not authorized to set the configuration
+  // - There already is a configuration, it just didn't appear in the index yet
+  // This method therefore does not try to re-submit the initial configuration in case of failure.
+  private def submitInitialConfig(writeService: state.WriteConfigService): Future[Unit] = {
+    val submissionId = submissionIdGenerator.generate()
+    withEnrichedLoggingContext("submissionId" -> submissionId) { implicit loggingContext =>
+      logger.info(s"No ledger configuration found, submitting an initial configuration.")
```

Comment:
I wish we had a lint option for catching these:
```suggestion
      logger.info("No ledger configuration found, submitting an initial configuration.")
```

Reply:
Caught it in #10461.